### PR TITLE
feat(composer)!: fallback to composer 2.2 LTS on constraint extraction

### PR DESCRIPTION
**Pull request by herndlm**
## Changes:

Let the constraint extraction fallback to Composer `^2.0` instead of `1.*`.

## Context:

With 2/3 of Composer installs being done via Composer 2 it should be fine to not fallback to 1 in case no constraint could be extracted. See https://github.com/composer/composer/issues/10340 and https://blog.packagist.com/deprecating-composer-1-support/
This should also improve the update behaviour in case no Composer dependency or lockfile exists to extract the composer constraint because #13154 can be made use of.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
Does composer V2 require a minimal PHP Version we can check?

File: lib/manager/composer/utils.ts
Comment (rarkins) [inline lib/manager/composer/utils.ts:69]: Is this the same as 2.*?

Comment (viceice) [reply lib/manager/composer/utils.ts:69]: No, if i remember right, star will match lowest and caret highest V2

**Comment by herndlm** [comment]
> Does composer V2 require a minimal PHP Version we can check?

2.2 runs still with 5.3+ just as v1 does (`^5.3.2 || ^7.0 || ^8.0` to be exact)
but 2.3 _will_ be 7.2+ (`^7.2.5 || ^8.0`)

**Comment by viceice** [comment]
Sounds good. So if an old PHP version is constraint, we should limit to `~2.2.0`

**Comment by herndlm** [comment]
adapted, like this, or did I misunderstand you? At first I thought of adding a default PHP constraint too, but I'm not sure what affects this will have in case no PHP version is required in composer.json.

**Review by viceice — DISMISSED** [review]

**Comment by herndlm** [comment]
had to update the artifacts snapshots after the recent changes and merges, overlooked that.

**Review by viceice — APPROVED** [review]

**Review by viceice — CHANGES_REQUESTED** [review]
I'm still not fully happy with this, as we probably break all existing v1 users.
I know composer v1 is now deprecated, but it will still be used by a lot of users and we won't like to break them.

@rarkins WDYT?

BTW: I've created #13359, because we need to migrate to pacvagist v2 API to find updates.

**Comment by herndlm** [comment]
> I'm still not fully happy with this, as we probably break all existing v1 users.
> I know composer v1 is now deprecated, but it will still be used by a lot of users and we won't like to break them.

Fyi I did a quick composer 1 update test (basically just regenerating the lockfile) in 2 apps. It was working just fine, with the following to note
- it was quite slow (apparently I'm already used to v2 too much. astonishing)
- Composer shows the warning `Warning from https://repo.packagist.org: Support for Composer 1 is deprecated and some packages will not be available. You should upgrade to Composer 2. See https://blog.packagist.com/deprecating-composer-1-support/`
- it changed things in the lockfile, uninteresting stuff, some keys here and there. and `plugin-api-version` was adapted of course. there are also some solver dependency changes although no packages were update for me. this could be related to some solver optimizations in v2.

Anyways, I'm not saying that this is compatible, but it should still work for most of the cases to just use another composer version. Using v2 for v1 projects means then that the lockfile is changed as well of course. Which might or might not be annoying I guess

**Comment by rarkins** [comment]
I would like to keep supporting Composer v1 if it's feasible. As a general rule, we don't "migrate" users from an older version of a package manager to a newer one by surprise if we can avoid it. e.g. in npm we try to detect if the older lock file is in use and use npm@6 if so and not npm@latest.

**Comment by herndlm** [comment]
OK this decision is of course fine with me. Shall I just close the PR then? Users like me that use 2(.2)+ mostly benefit from its features already if a lockfile is there (= the last use composer version can be extracted as constraint). In case without a logfile it might not be able to extract the constraint and use the newer features I guess. But as workaround I can just add that constraint to my composer.json

I mostly thought that the Renovate (full) docker image is already shipping with composer 2 anyways and therefore it makes sense to fall back to using all it's features.

**Comment by viceice** [comment]
Contraints are only used if you run `binarySource=docker`, so the full image is always using the (latest) global composer.

Some managers also support containts when  `binarySource=install`.

**Comment by herndlm** [comment]
> Contraints are only used if you run `binarySource=docker`, so the full image is always using the (latest) global composer.
> 
> Some managers also support containts when `binarySource=install`.

Right, I understand that. Renovate should then still use Composer 1 correctly if it can extract the constraint (= a lockfile or direct composer dependency is there), right?
But I don't want to start a discussion here and annoy both of you. Feel free to just close this PR then, if you think it's still too early :)

**Review by rarkins — DISMISSED** [review]

**Comment by rarkins** [comment]
@viceice I'm ok with falling back to latest composer rather than 1.*, if no other constraint is found

**Review by viceice — COMMENTED** [review]
Ok, so we break V1 users now. Should we classify as major update?

**Comment by rarkins** [comment]
Aren't v2 users "broken" now instead?

And aren't v1 users *not* broken if we detect a constraint through our usual methods? i.e. they're broken only if they use v1 AND we can't detect that?

**Comment by rarkins** [comment]
I'm not sure if this is breaking or not, but let's just wait for the next major and note that Composer will default to 2.2+ if no constraints are detected

**Comment by viceice** [comment]
I think composer V2 will always add something to the lockfile, so we can detect it as long as there is a lockfile

**Comment by rarkins** [comment]
So you prefer to detect v2 and keep falling back to v1?

**Comment by viceice** [comment]
Currently yes, i but we can of cause change this in a major update.

Maybe wait until composer V1 is fully EOL?

**Comment by rarkins** [comment]
@herndlm Did you have a use case which wasn't currently met?

**Comment by herndlm** [comment]
It was mostly about repos without a checked-in lockfile, where composer version detection is not optimal I would say. The workaround is to require a specific composer (plugin) api though, which might be fine too.

**Comment by viceice** [comment]
Isn't it best practices to checkin the lockfile?

**Comment by rarkins** [comment]
Is there any time we'd run composer if no lock file detected?

**Comment by herndlm** [comment]
> Isn't it best practices to checkin the lockfile?

For apps - definitely! For libs it's hard to say. There's no real consense in the community I think, some of the dependency "hardliners" check it im, most of the people don't. Which is kind of confirmed by https://getcomposer.org/doc/01-basic-usage.md#commit-your-composer-lock-file-to-version-control and the fact that the composer maintainers themselves don't check it in for libs. There are definitely up and downsides to this.


> Is there any time we'd run composer if no lock file detected?

not sure if I understood you right, but it would still run and update requirement constraints in composer.json?

**Comment by rarkins** [comment]
We don't use `composer` the binary in order to update `composer.json`. Going to close in this case

**Comment by herndlm** [comment]
> We don't use `composer` the binary in order to update `composer.json`. Going to close in this case

ah I get you now. You're absolutely right. Then my use cases are covered by having a lockfile.

Thx and sorry for the noise/misunderstanding!

**Pull request closed without merging**
